### PR TITLE
MultiHeadAttention Layer Bug Fix

### PR DESCRIPTION
**Pull request by kumarutkarsh1248**
This PR address [issue](https://github.com/mlpack/mlpack/issues/3969)

Fixed the forward pass implementation in the MultiHeadAttention layer to produce the correct output.
Made relevant changes in the tests, except for the gradient tests.

Next Steps:
Get feedback on the changes and extend the fixes to the backward pass and gradient computation.

### CODE REVIEW

**Comment by rcurtin** [comment]
I pushed changes for the backward and gradient pass.  While debugging those, I realized that the forward pass on the softmax didn't work correctly when the masks used the intended values of `-Inf`.  So, I implemented a custom forward masked softmax pass manually.  It's not particularly pretty but it works correctly.

I believe there are a ton of optimizations that can be made to this implementation (mostly in memory allocations), but we can return to that later---correctness is the important part here.

**Comment by kumarutkarsh1248** [comment]
Using `std::numeric_limits::lowest() `is actually risky, I just realized. When I was testing the forward pass locally, I set the attention mask to all zeros, so I didn’t notice this problem at the time.
Do you think there’s any way we can still use `softmax.forward()` later when we optimize the implementation?

**Comment by rcurtin** [comment]
Yeah, the idea was to actually use `-Inf` as the masking value, not `-DBL_MAX`.

I'm not sure if using `softmax.Forward()` will be possible---I think the implementation here will change a lot when it gets optimized.  We'll see?
